Optimize and re-register regexp_replace Spark function

codyschierbeck opened the pull request:
Addresses #7856 and #8438 

https://github.com/facebookincubator/velox/pull/8283 enables reintroducing SparkSQL regexp_replace.

Re-registers regexp_replace, optimizes its call function for performance and readability, and removes regexp_replace from Spark fuzzer.

regexp_replace previously caused fuzzer failures due to the interaction between dictionary encoding, common vs simple expression evaluation, and the regexp_replace cache limit exception. This interaction would lead to an exception being thrown at separate points in evaluation, and break the fuzzer.

codyschierbeck commented:
**Fuzzer details:**

I ran the spark_expression_fuzzer_test executable with the following command

`spark_expression_fuzzer_test --duration_sec 1800 --enable_variadic_signatures --lazy_vector_generation_ratio 0.2 --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --max_expression_trees_per_step 2 --retry_with_try --enable_dereference --seed 3732405514 --logtostderr=1 --minloglevel=0 --assign_function_tickets regexp_replace=10000`

Relevant lines from output are: 

```
I0110 12:43:05.538235 2318452 ExpressionFuzzerVerifier.cpp:363] ==============================> Done with iteration 1869
I0110 12:43:05.538623 2318452 ExpressionFuzzerVerifier.cpp:160] ==============================> Top 10 by number of rows processed
I0110 12:43:05.538632 2318452 ExpressionFuzzerVerifier.cpp:162] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0110 12:43:05.538637 2318452 ExpressionFuzzerVerifier.cpp:166] regexp_replace 13375 51.50% 315623

..............


E0110 12:43:05.539583 2318452 ExpressionFuzzerVerifier.cpp:369] Total iterations: 1870
E0110 12:43:05.539592 2318452 ExpressionFuzzerVerifier.cpp:370] Total failed: 598
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

In debugging the fuzzer to try and figure out why regexp_replace was causing issues, I ran into some inconveniences.  Oftentimes the expression that caused problems occured after many hundreds of regexp_replace calls. Hitting a breakpoint hundreds of times for each iteration of debugging was not ideal. I tried leveraging velox/expressions/tests/velox_expression_runner_test to skip to the call that broke the fuzzer, but I found that the velox_expression_runner library was only linked to velox_functions_prestosql.

```
add_library(velox_expression_runner ExpressionRunner.cpp)
target_link_libraries(velox_expression_runner velox_expression_verifier
                      velox_functions_prestosql velox_parse_parser gtest)
```

To get around this, I made my own spark_expression_runner and spark_expression_runner_test to help with debugging. 

```
add_library(spark_expression_runner ExpressionRunner.cpp)
target_link_libraries(spark_expression_runner velox_expression_verifier
                      velox_functions_spark velox_parse_parser gtest)


add_executable(spark_expression_runner_test SparkExpressionRunnerTest.cpp)
target_link_libraries(
  spark_expression_runner_test spark_expression_runner velox_exec_test_lib
  velox_function_registry gtest gtest_main)
```

You can find this at: https://github.com/codyschierbeck/velox/commit/117508c5d9ff383138009134155e8b06501207d0

@mbasmanova if you're interested, I can make a pull request to integrate these changes. I also have a version that instead of making a unique expression runner for Spark I just linked velox_functions_spark to the velox_expression_runner. It worked and built, but I am sure there are issues with matching function registration that I just managed to not encounter in my debugging.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -107,63 +107,111 @@ void ensureRegexIsConstantAndCompatible(
+    try {
+      VELOX_USER_CHECK_GT(
+          *position, 0, "regexp_replace requires a position >= 1");
+    } catch (const VeloxUserError& e) {
```
This logic doesn't make sense to me. Why are you doing it? You should be able to use VELOX_USER_CHECK_GT as is without try-catch-rethrow.

CC: @laithsakka @kgpai

mbasmanova commented:
> To get around this, I made my own spark_expression_runner and spark_expression_runner_test to help with debugging.

This sounds useful. Would you like to submit a PR? CC: @kgpai @kagamiori

mbasmanova commented:
> Second, it threw exceptions without ensuring that the result was set to null.

This should be fine. It is Ok to throw from a function without setting result to anything particular. Perhaps, the problem is somewhere else.

CC: @kevinwilfong @bikramSingh91 @kgpai @kagamiori @laithsakka

codyschierbeck commented:
**Macro Details**

@mbasmanova Wrote this out before you commented, was about to press send. In my testing with the fuzzer I found that if I threw an exception without setting result to null (result.resize(0)), the fuzzer occasionally failed. Below is a more detailed explanation. 

**Written before your comment**

I feel that this pull request is ready for review, but before I change it to that I want discuss the workaround I used for setting the result to null before throwing an exception. Example:

```cpp
    try {
      VELOX_USER_CHECK_GT(
          *position, 0, "regexp_replace requires a position >= 1");
    } catch (const VeloxUserError& e) {
      result.resize(0);
      throw e;
    }
```

The catch block in this case is really just a callback function for after the CHECK_GT check fails. I think try/catch block is fine, if not bulky, but I wanted to see if there was a way to keep the VELOX_USER_CHECK_GT as one line. 

To test, I defined the following macro in RegexReplace.cpp

```cpp
#define VELOX_USER_CHECK_GT_CB(lhs, rhs, callback) \
    if (!((lhs) > (rhs))) { \
        callback; \
        VELOX_USER_FAIL("Check failed: " #lhs " > " #rhs); \
    }

```

Then, for debugging purposes, I created  a wrapper for result.resize(0) so I could place a breakpoint on it.

```cpp

  void wrapSetNull(out_type<Varchar>& result) {
    result.resize(0);
  }
```

With these two features in place, the try/catch block becomes

```cpp
VELOX_USER_CHECK_GT_CB(*position, 0, wrapSetNull(result));

```

Testing with my debugger showed that the callback function does indeed get called. And after running velox/functions/sparksql/tests/velox_functions_spark_test, all tests still passed. 

**Written after your comment**

After seeing your comment, I ran grep -rnw ./velox -e "bool call" to try and find any other cases where a VELOX_USER_CHECK or any other macro from that family was called. I found a couple cases, like in velox/functions/prestosql/Bitwise.h  BitwiseArithmeticShiftRightFunction. 

I remember running into that fuzzer issue weeks ago and made a note in my head to incorporate that fix after the fuzzer override args issues were resolved, so I will try and replicate now to make sure there wasn't something else going on. 

If I fail to replicate, you can just ignore the callback macro idea.

codyschierbeck commented:
@mbasmanova 

I reverted to the original macro calls with no try-catch-rethrow and ran 3 parallel instances of the fuzzer with 10,000 regexp_replace tickets for 30 minutes each. All three succeeded without failure so the fuzzer issue I ran into before must have been caused by something else (though it is weird that at the time try-catch-rethrow fixed it). Pushing those changes then switching to ready for review. 

Drafting a PR for spark_expression_runner_test and gathering some usage examples as well.

mbasmanova commented:
@codyschierbeck Thanks. Would you update PR description?

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -107,63 +107,91 @@ void ensureRegexIsConstantAndCompatible(
+    if (!stringInput) {
+      result.resize(0);
+      return false;
+    } else if (!pattern || !replace) {
```
nit: drop else after return for readability

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -107,63 +107,91 @@ void ensureRegexIsConstantAndCompatible(
+      const arg_type<Varchar>* pattern,
+      const arg_type<Varchar>* replace) {
+    if (!stringInput) {
+      result.resize(0);
```
not needed since you return a null

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -107,63 +107,91 @@ void ensureRegexIsConstantAndCompatible(
+      std::memcpy(result.data(), stringInput->data(), stringInput->size());
+      return true;
+    }
+    re2::RE2* patternRegex = nullptr;
```
combine this line with next

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -107,63 +107,91 @@ void ensureRegexIsConstantAndCompatible(
+      const arg_type<Varchar>* replace,
+      const arg_type<int64_t>* position) {
+    if (!position) {
+      return callNullable(result, stringInput, pattern, replace);
```
Is there a lot of copy-paste between two functions: with and w/o position argument? Couldn't function w/o position call the one with position using position value 1 or similar?

mbasmanova reviewed (commented):
@codyschierbeck Thank you for the fixes. It is a bit hard to read the changes. Would it be possible to extract function rename into a separate PR?

codyschierbeck commented:
@mbasmanova 

Changed description. Working on the changes that you have suggested as well. 

As for splitting into two PRs, is there a way to make one pull request dependent on another in Github? i.e. Make the name change PR and then make a PR for callNullable that depends on name change? I wanted to do that with this PR while waiting for https://github.com/facebookincubator/velox/pull/8283 but I didn't think it was doable. 

If not, would the right practice be to create both PRs, and when the name change is merged rebase the second?

mbasmanova commented:
> is there a way to make one pull request dependent on another in Github?

I don't think so.

> If not, would the right practice be to create both PRs, and when the name change is merged rebase the second?

This would be the way. Thanks.

codyschierbeck commented:
@mbasmanova 

Hi, I believe this is ready for another review. 

I implemented the changes you requested and removed duplicate code. The 3 args version now just calls the 4 args with position as a nullptr. The 4 args already needed to support 'position' being null, so when it is null it now sets 'positionVal' to 1.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -110,60 +110,65 @@ template <typename T>
-    } else {
-      result.resize(0);
-    }
+      const arg_type<Varchar>* stringInput,
```
Would you check if documentation needs updating to clarify the behavior for input NULLs and explain when result might be NULL?

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -110,60 +110,65 @@ template <typename T>
+    if (position) {
+      positionVal = *position;
+    }
+    if (!stringInput) {
```
Move this before computing positionVal.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -110,60 +110,65 @@ template <typename T>
+      return false;
+    }
+    if (!pattern || !replace) {
+      result.resize(stringInput->size());
```
Can you not do result = stringInput? CC: @kagamiori @laithsakka

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -110,60 +110,65 @@ template <typename T>
+    }
+
+    VELOX_USER_CHECK_GT(
+        positionVal, 0, "regexp_replace requires a position >= 1");
```
Use _GE to match ">= 1" in the message

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -260,6 +260,7 @@ TEST_F(RegexFunctionsTest, regexpReplaceSimplePosition) {
   std::string output = "Hello WorLd";
   auto result = testRegexpReplace("Hello World", "l", "L", {6});
   EXPECT_EQ(result, output);
+  EXPECT_THROW(testRegexpReplace("Hello World", "l", "L", {0}), VeloxUserError);
```
Are you missing test cases for the new patterns being supported?

Use VELOX_ASSERT_THROW

codyschierbeck commented:
Hi, @mbasmanova. Apologies for the long time between replies. I discovered some issues with RegexpReplace while trying to implement some changes I felt were necessary. When you get a chance, please check out #8438 which goes into detail about these changes. 

Once you've taken a look, I have the relevant solution to those problems ready to be pushed to this PR. I believe I will have to change the PR name to "Charge SparkSQL regexp_replace to a 'bool call' UDF" and have the description reflect that.

codyschierbeck commented:
@kgpai @mbasmanova I am seeing build errors with macOS. If I'm reading this correctly there was something wrong with building rust?

```
2024-01-19T10:29:01.7392360Z [32m==>[0m [1mInstalling ccache dependency: [32mrust[39m[0m
2024-01-19T10:33:52.7130080Z [34m==>[0m [1mPatching[0m
2024-01-19T10:33:53.5803550Z [34m==>[0m [1m./install.sh --prefix=/private/tmp/rust-20240119-12605-9c7msu/rustc-1.75.0-s[0m
2024-01-19T10:34:01.3585310Z [34m==>[0m [1m./configure --sysconfdir=/usr/local/etc --tools=analysis,cargo,clippy,rustdo[0m
2024-01-19T10:34:02.9833430Z [34m==>[0m [1mmake[0m
2024-01-19T11:19:57.5984730Z ##[error]The operation was canceled.
2024-01-19T11:19:57.6203070Z Post job cleanup.
```

The output isn't very verbose.

kgpai commented:
@codyschierbeck Seems to be some dependency issue, unrelated to your changes. I am rerunning the job , I will investigate the failure independent of your change though and it shouldnt block this pr.

mbasmanova commented:
@codyschierbeck I commented on https://github.com/facebookincubator/velox/issues/8438#issuecomment-1900234319 a few days ago.

codyschierbeck commented:
> @codyschierbeck I commented on [#8438 (comment)](https://github.com/facebookincubator/velox/issues/8438#issuecomment-1900234319) a few days ago.

Apologies, I will take a look

codyschierbeck commented:
@mbasmanova I just submitted my reply on the issue with my best attempt at explaining the corner case and addressing your concerns.

PHILO-HE commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -165,6 +170,24 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+    return result;
+  }
+
+  int randomPosition(std::mt19937& generator) {
```
Do we need this function & the above one? I didn't find where they are used.

codyschierbeck commented:
@kgpai RuntimeMetrics.cpp line 17 is causing build issues with linux-adapters for both of my recently updated PRs, this and #8341 . Seems to be an issue with fmt update?
```
/usr/local/include/fmt/core.h:1580:7: error: static assertion failed: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
 1580 |       formattable,
      |       ^~~~~~~~~~~
 ```

kgpai commented:
@codyschierbeck I dont see the full compile logs . Whats the type its complaining about ? you can add a specialization like something below in the failing file : 

```
template <>
struct fmt::formatter<YOUR_TYPE> : formatter<int> {
  auto format(YOUR_TYPE s, format_context& ctx) {
    return formatter<int>::format(static_cast<int>(s), ctx);
  }
};
```

kgpai commented:
@codyschierbeck I suspect this problem is because the containers have been updated to use fmt 9 but this branch doesnt have the latest changes. Can you rebase against main and see if that fixes it.

codyschierbeck commented:
@mbasmanova @kgpai 

I believe this is ready. 

To summarize:

- we are re-registering and modifying regexp_replace to return bool instead of null, since throwing errors returns a null value. 
- The 3 arg version of the function now calls the 4 arg version with its default position value of 1. 
- Extracted calculating UTF-8 position to its a private function for readability.
- Moved this function call before getRegex function call to avoid compiling a regex needlessly if getUTF8Positon throws an exception.
- Instead of using result.resize() and std::memcpy to set result, we leverage the StringView and StringWriter APIs.
- Added regexp_replace to spark_expression_fuzzer_test skipList.
- Included @PHILO-HE changes from #8387 for integers and lower case
- Modified test cases that EXPECT_THROW to ASSERT the exact expected messages
- Removed as many intermediate variables as possible, the remaining intermediate values are:


1. replaceStringPiece and patternRegex, needed for GlobalReplace API
2. prefix and targetString. RE2::GlobalReplace does not support a position argument, so instead we run it on everything that follows the prefix.

mbasmanova reviewed (commented):
@codyschierbeck Would you double check PR description to make sure it matches the code changes? Also, do we need to update the function documentation to explain cases when it returns NULL?

codyschierbeck commented:
> @codyschierbeck Would you double check PR description to make sure it matches the code changes? Also, do we need to update the function documentation to explain cases when it returns NULL?

Okay, working on this now.

codyschierbeck commented:
@mbasmanova I changed the description and updated the documentation. If you'd like, I can squash commits since we are up to 11 now.

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -38,7 +38,9 @@ See https://github.com/google/re2/wiki/Syntax for more information.
 
     Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite``. If no match is found, the original string is returned as is.
-    There is a limit to the number of unique regexes to be compiled per function call, which is 20.
+    There is a limit to the number of unique regexes to be compiled per function call, which is 20. If this limit is exceeded the function will throw an exception and return null values.
```
> If this limit is exceeded the function will throw an exception and return null values.

A function cannot both throw an exception and return null value. I think this should say " throw an exception".

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -38,7 +38,9 @@ See https://github.com/google/re2/wiki/Syntax for more information.
     Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite``. If no match is found, the original string is returned as is.
-    There is a limit to the number of unique regexes to be compiled per function call, which is 20.
+    There is a limit to the number of unique regexes to be compiled per function call, which is 20. If this limit is exceeded the function will throw an exception and return null values.
+    regexp_replace can also return null values if the any provided argument is null, ``string`` contains an invalid UTF-8 character, or ``pattern`` does not conform to RE2 syntax: https://github.com/google/re2/wiki/Syntax.
```
"if the any" -> "if any"

> or ``pattern`` does not conform to RE2 syntax

Does Spark also return NULL if regexp is invalid?

mbasmanova replied:
Shouldn't Velox implementation also throw in these cases?

codyschierbeck replied:
The only times this function returns 'null' is when an exception is thrown. Now that I am thinking about it, I originally believed the fuzzer errors I was running into were because I threw exceptions in a 'void call' function, but now that we figured out it was due to cache + dictionary encoding, I'm questioning my understanding of bool vs void call.

If the only times this function 'returns' null is after throwing an exception, would void call be the correct implementation? If that's the case, it would be a really quick fix to implement. I am currently building that version to test the fuzzer again

mbasmanova replied:
@codyschierbeck 'void call' is appropriate if function either completes successfully and produces a non-null result or fails. 'bool call' is needed only if function may complete successfully and produce a null result.

mbasmanova replied:
> Would "Optimize and re-regeister SparkSQL regexp_replace" be a better name for this PR?

That would be great. 

P.S. I'm so impressed with your ability to produce lengthly clear detailed writings in no time at all. Sorry that I cannot match this beautiful communication style in my terse responses.

codyschierbeck replied:
Thank you very much, that honestly made my day.

I will say that it is because I genuinely enjoy working on Velox. I concentrated my CS Master's in Big Data, but was never given an opportunity at work to leverage that until being assigned to Velox. So I was super excited to get started even though I knew practically zero C++ before diving head-first into this codebase. 

I'm probably not the best judge of this, but it is such an impressive codebase. It has been a fantastic in helping me learn. For example, today I was trying to learn how I might go about implementing either a Tuple or VariadicTuple type in case that is what the SpecialForm of the decode signature ends up requiring. My debugger ended up leading me into the UdfTypeResolver.h which made me look at Type.h, VELOX_FUNCTION_DEFINE_TYPES(T), and basically the entire framework in a whole new light.

That feeling happens pretty much every time I face a new problem with this codebase, so even though at times I've been pretty embarrassed for how obvious the solutions end up being, I enjoy it.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -301,15 +306,21 @@ TEST_F(RegexFunctionsTest, regexpReplaceWithEmptyString) {
-  EXPECT_THROW(testRegexpReplace("[]", "[a&&[^b]]", ""), VeloxUserError);
+  VELOX_ASSERT_THROW(
+      testRegexpReplace("[]", "[a[b]]", ""),
+      "regexp_replace does not support character class union, intersection, or difference ([a[b]], [a&&[b]], [a&&[^b]])");
```
This test doesn't match documentation which says that the function returns NULL in this case.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -301,15 +306,21 @@ TEST_F(RegexFunctionsTest, regexpReplaceWithEmptyString) {
+  VELOX_ASSERT_THROW(
       testRegexpReplace(std::string("\xA0") + "bcacbdefg", "", "", {2}),
-      VeloxUserError);
+      "regexp_replace encountered invalid UTF-8 character");
```
The docs say the function returns NULL in this case.

mbasmanova commented:
> I changed the description and updated the documentation. 

Thank you.

> I can squash commits since we are up to 11 now.

Up to you. I noticed GitHub slows down when there are many commits.

codyschierbeck commented:
@mbasmanova I made those changes, squashed, and pushed. 

I'm seeing something weird with the documentation diff right now. I'm not confident that they will compile well based on the eye test, but I deleted conda off my dev machine because it was taking priority over my other python environments and prioritizing its environment variables. I will have to re-install that more carefully tomorrow morning to make sure they do compile as expected.

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -57,8 +59,10 @@ See https://github.com/google/re2/wiki/Syntax for more information.
-    There is a limit to the number of unique regexes to be compiled per function call, which is 20.
+    Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite`` starting from the specified ``position``.  If no match is found, the original string is returned as is. If the ``position`` is less than one, the function throws an exception. If ``position`` is greater than the length of ``string``, the function returns the original ``string`` wi ...
+    There is a limit to the number of unique regexes to be compiled per function call, which is 20. If this limit is exceeded the function will throw an exception.
+    regexp_replace will throw an exception if any provided arguments are null, if ``string`` contains an invalid UTF-8 character, if ``position`` is less than 1, or if ``pattern`` does not conform to RE2 syntax: https://github.com/google/re2/wiki/Syntax.
```
> regexp_replace will throw an exception if any provided arguments are null

This is surprising behavior. Does it match Spark?

PHILO-HE replied:
Just confirmed. This function has default null behavior in Spark, i.e., returns null for any null input. I note the null behavior is already aligned with this Spark. @codyschierbeck, please correct the doc. Thanks!

codyschierbeck replied:
Apologies, this was doc mistake when trying to remove all the references to returning null that should have been throwing an exception. Adjusting.

mbasmanova reviewed (commented):
Looks great % a few nits.

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -135,79 +125,76 @@ struct RegexpReplaceFunction {
+    // Adjust the position for UTF-8 by counting the code points.
+    size_t utf8Position = 0;
+    size_t numCodePoints = 0;
+    while (numCodePoints < max_length - 1 && utf8Position <= len) {
```
If I am reading that correctly, I believe so. I would call it as follows

```cpp
size_t utf8Position = 
facebook::velox::functions::stringImpl::cappedLength::cappedLength<false>(stringInput, position);
```

I have two concerns and one idea for readability:

1. My implementation throws when it encounters a bad byte, cappedLengthUnicode skips it. This should cause only regexpReplaceUnvalidUTF8 to fail in testing.
2. My implementation is aware of the '1-indexed' condition, whereas cappedLength appears to be 0-indexed. 
3. The call has a pretty long namespace. For readability, just make getUTFLength an encapsulation of cappedLength so the code is still clear about what it is doing. If you feel this is unnecessary, I can just make the full call in a single line.

While writing this up I built and tested, the following test cases now fail:

```
[  FAILED  ] 16 tests, listed below:
[  FAILED  ] RegexFunctionsTest.regexpReplaceRegistration
[  FAILED  ] RegexFunctionsTest.regexpReplaceNonAsciiPositionNoChange
[  FAILED  ] RegexFunctionsTest.regexpReplaceMatchSparkSqlTestSimple
[  FAILED  ] RegexFunctionsTest.regexpReplaceWithEmptyString
[  FAILED  ] RegexFunctionsTest.regexpReplaceInvalidUTF8
[  FAILED  ] RegexFunctionsTest.regexpReplacePosition
[  FAILED  ] RegexFunctionsTest.regexpReplaceMultipleMatches
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframe
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframeMultiple
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframeSpecial
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframeSizes
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframeCharacterTypes
[  FAILED  ] RegexFunctionsTest.regexpReplaceMatchSparkSqlTest
[  FAILED  ] RegexFunctionsTest.regexpReplaceRowsNoPosition
[  FAILED  ] RegexFunctionsTest.regexpReplaceMassiveVectors
[  FAILED  ] RegexFunctionsTest.regexpReplaceCacheMissLimit

```
Passing position - 1 and adding 1 to the output does not resolve any test cases, in fact an additional test case fails. This isn't necessarily because of the new function, it could be how I am calling it, so I will take a deeper look and gather details.

codyschierbeck replied:
Adjusting the call to use "position - 1" and not add 1 back to the end was the correct call. It now only fails the following:

```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] RegexFunctionsTest.regexpReplaceNonAsciiPositionNoChange
[  FAILED  ] RegexFunctionsTest.regexpReplaceInvalidUTF8
```

I'll look into how to fix the first since we can just remove or adjust the second. 

Is skipping bad bytes the standard way of handling them? I'm re-rewriting the InvalidUTF8 case to instead expect some sort of valid output, but I am not sure if this is the right idea.

codyschierbeck replied:
@mbasmanova I don't think we can replace getUTF8Positon with cappedLength. 

cappedLength returns the number of characters it encounters:
```cpp
FOLLY_ALWAYS_INLINE int64_t
cappedLengthUnicode(const char* input, size_t size, size_t maxChars) {
  // First address after the last byte in the input
  auto end = input + size;
  auto currentChar = input;
  int64_t numChars = 0;

  // Use maxChars to early stop to avoid calculating the whole
  // length of long string.
  while (currentChar < end && numChars < maxChars) {
    auto charSize = utf8proc_char_length(currentChar);
    // Skip bad byte if we get utf length < 0.
    currentChar += UNLIKELY(charSize < 0) ? 1 : charSize;
    numChars++;
  }

  return numChars;
}
```

whereas getUTF8Length returns the sum of the number of bytes it has encountered

```cpp

  size_t getUTFLength(const char* str, size_t len, size_t maxCharacters) {
    // Adjust the position for UTF-8 by counting the code points.
    size_t utf8Position = 0;
    size_t numCodePoints = 0;
    int64_t numChars = 0;
    auto currentChar = str;
    auto end = str + len;
    while (numCodePoints < maxCharacters - 1 && utf8Position <= len) {
      int charLength = utf8proc_char_length(str + utf8Position);
      VELOX_USER_CHECK_GT(
          charLength, 0, "regexp_replace encountered invalid UTF-8 character");
      ++numCodePoints;
      utf8Position += charLength;
    }
```

So, if we encounter multi-byte characters like the 3 byte ♫ or the 2 byte ¡, the return values differ between the two functions.

getUTFLength's return value is used to split the string between prefix and target, so those extra bytes make a difference in some cases.

codyschierbeck commented:
@mbasmanova Implemented your review requests:

1. Removed unnecessary return in 3 arg call
2. Moved utfPositon > stringInput.size() check before getRegex to avoid uselessly compiling regexes
3. Moved getUTFLength definition above member variable 'cache_'
4. Adjusted getUTFLength's signature and behavior to be more general. Now takes a string-like input and gets the data and length using .data() and .size() API. Changed its local variable 'len' to 'bytes' to be more clear about what it represents. maxCharacters is now assumed to be correct and we now check for numCodePoints < maxCharacters instead of maxCharacters - 1. The call to this function in 'void call' now adjusts the positon itself.
5. Adjusted documentation to no longer claim an exception is throw when NULL is passed as an argument
6. Adjusted an example in the 4 arg documentation to no longer claim a negative position will return the input string. Made the value positive
7. Added spacing to both documentation's as without the extra spaces were large and hard on the eyes.

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -74,6 +84,6 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+        SELECT regexp_replace('Hello, World!', 'l', 'L', 5); -- 'Hello, World!'
 
-        SELECT regexp_replace('Hello, World!', 'l', 'L', 100); -- ERROR: Position exceeds string length.
+        SELECT regexp_replace('Hello, World!', 'l', 'L', 100); -- 'Hello, World!
```
> 'Hello, World!

Are we missing closing quote?

codyschierbeck replied:
Sorry to hear that weren't feeling okay, I hope you're doing better now.

And yes, we are. I will make sure this is fixed.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -135,79 +124,76 @@ struct RegexpReplaceFunction {
+    return utf8Position;
+  }
+
+  mutable folly::F14FastMap<std::string, std::unique_ptr<re2::RE2>> cache_;
```
Does this need to be 'mutable'?

codyschierbeck replied:
Tested locally and removed mutable. Got the following compile error:


`/velox/velox/functions/sparksql/RegexFunctions.cpp:176:5: error: cannot convert ‘const folly::F14FastMap<std::__cxx11::basic_string<char>, std::unique_ptr<re2::RE2> >*’ to ‘folly::f14::detail::F14BasicMap<folly::f14::detail::VectorContainerPolicy<std::__cxx11::basic_string<char>, std::unique_ptr<re2::RE2>, void, void, void, std::integral_constant<bool, true> > >* 
176 |     cache_.emplace(pattern, std::move(patternRegex));`

The mutable keyword is still new to me but after looking online, I believe this error occured because getRegex was a constant function, so to modify the member "cache_" of RegexpReplaceFunction within it, cache_ needed to be mutable.

 I don't recall why I made it constant (it could have been to resolve the inverse of this error), and I don't know the implications of removing it. That said everything compiles and all tests pass when you remove mutable from cache_ and const from getRegex. If there's nothing setting off alarms in your head with these changes, they will be included in the next push later tonight.

mbasmanova replied:
> remove mutable from cache_ and const from getRegex

Let's do that. Thanks.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -135,79 +124,76 @@ struct RegexpReplaceFunction {
-  mutable folly::F14FastMap<std::string, std::unique_ptr<re2::RE2>>
-      patternCache_;
+  template <typename TString>
+  size_t getUTFLength(TString& input, size_t maxCharacters) {
```
It would be nice to move this function next to other UTF-8 functions, document it and add a dedicated unit test.

can input be a const &?

codyschierbeck replied:
Tested locally with const TString& input, everything compiled and all tests passed.

To clarify, do you mean moving the function next to other UTF-8 functions in velox/functions/lib/string/StringImpl.h?

I think that would be pretty straightforward and I can include it later tonight. Do you feel that would be better as its own PR since it is out of the scope of SparkSQL regexp_replace? I am happy with either solution.

In that context, would a better name for this function be "cappedUTFBytes/cappedUnicodeBytes"?

mbasmanova replied:
I really love the way you go extra mile and really explore the space of possibilities. Thank you for doing this.

I'm thinking that this function could go next to cappedLength and perhaps, it can be named cappedByteLength to allow for distinguishing between length in characters vs. length in bytes. It can be part of this PR.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -529,24 +544,6 @@ TEST_F(RegexFunctionsTest, regexpReplaceMassiveVectors) {
   assertEqualVectors(result, output);
 }
 
-TEST_F(RegexFunctionsTest, regexpReplaceCacheLimitTest) {
```
Why are you removing this test?

codyschierbeck replied:
I think is just an artifact of my initial removal of the cache limit. Will add back in push later tonight.

mbasmanova reviewed (commented):
@codyschierbeck Sorry for the delay. I wasn't feeling well.

This is shaping up nicely. Looks good overall % a few comments.

mbasmanova commented on velox/expression/tests/SparkExpressionFuzzerTest.cpp:
```diff
@@ -47,6 +47,7 @@ int main(int argc, char** argv) {
   // rlike, md5 and upper
   std::unordered_set<std::string> skipFunctions = {
       "regexp_extract",
+      "regexp_replace",
```
nit: would you add a comment with a link to GitHub issue where we have discussed why we have to skip this function?

mbasmanova commented on velox/functions/lib/string/StringImpl.h:
```diff
@@ -122,6 +122,25 @@ FOLLY_ALWAYS_INLINE int64_t cappedLength(const T& input, size_t maxLength) {
   }
 }
 
+/// Return a capped bytes(controlled by maxCharacters) of a string.
```
nit: a capped length in bytes (...)

and modify the comment on L114: a capped length in characters (...)

mbasmanova commented on velox/functions/lib/string/StringImpl.h:
```diff
@@ -122,6 +122,25 @@ FOLLY_ALWAYS_INLINE int64_t cappedLength(const T& input, size_t maxLength) {
+/// are multi-byte characters present in the input string.
+template <typename TString>
+FOLLY_ALWAYS_INLINE size_t
+cappedUnicodeBytes(TString& input, size_t size, size_t maxCharacters) {
```
nit: perhaps, cappedByteLength

const & for input

mbasmanova replied:
Why is 'size'? Is it needed? Can we not use input.size()?

mbasmanova commented on velox/functions/lib/string/StringImpl.h:
```diff
@@ -122,6 +122,25 @@ FOLLY_ALWAYS_INLINE int64_t cappedLength(const T& input, size_t maxLength) {
+  const char* str = input.data();
+  size_t utf8Position = 0;
+  size_t numCharacters = 0;
+  while (numCharacters < maxCharacters && utf8Position < size) {
```
I wonder if it makes sense to change cappedLengthUnicode to return a pair of numbers: length in bytes + length in characters, then re-use it here.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -135,79 +125,62 @@ struct RegexpReplaceFunction {
     }
+
+    size_t utf8Position =
+        facebook::velox::functions::stringImpl::cappedUnicodeBytes(
```
drop 'facebook::velox::'

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -135,79 +125,62 @@ struct RegexpReplaceFunction {
+      }
     }
+
+    size_t utf8Position =
```
nit: utf8Position -> start

If you look at usages of cappedLength, you'll notice that it has a template parameter isAscii. This is because ASCII strings are a lot cheaper to process than UTF-8 and there are a lot of ASCII-only datasets.

I suggest to model this function the same way as cappedLength and add callAscii here. See https://facebookincubator.github.io/velox/develop/scalar-functions.html#all-ascii-fast-path

mbasmanova reviewed (commented):
Looks good % some comments. It does look like it would be better to extract cappedByteLength into a separate PR though.

codyschierbeck commented:
@mbasmanova Hi, working on an initial draft PR for the cappedByteLength right now. Should be up within the next hour or so. After that, I will return to addressing your suggestions here

codyschierbeck commented:
@mbasmanova Implemented review requests

### Summary

1. Implemented callAscii for 3 and 4 arg values. Extracted duplicate code from UTF and ASCII implementations into private functions.
2. Removed InvalidUTF8 test case, further testing revealed Spark handles this use case without throwing exceptions
3. Modified the placeholder `cappedUnicodeBytes` to no longer throw an exception, use const input, removed size parameter, and changed call to use only the namespace `functions::stringImpl` instead of the full `facebook::velox...`. 
4. Added a reference to regexp_replace issue in SparkExpressionFuzzer to explain its inclusion in skipList
5. Renamed UTF8Position to `start` for clarity

### Details

I implemented callAscii for both signatures. This lead to a lot of duplicated code as the only difference between the two charsets execution is the UTF8Position value. So I extracted all the checks that occur before it into "performChecks" private function and the replacement code that occur after it into "performReplace". The top level call for callAscii with 4 paramaters looks like the following now: 
```cpp
  void callAscii(
      out_type<Varchar>& result,
      const arg_type<Varchar>& stringInput,
      const arg_type<Varchar>& pattern,
      const arg_type<Varchar>& replace,
      const arg_type<int64_t>& position) {
    if (performChecks(result, stringInput, pattern, replace, position - 1)) {
      return;
    }
    performReplace(result, stringInput, pattern, replace, position - 1);
  }
```
 where perform checks returns true if we detected that we can just pass the inputString or replace into result. 

Top-level call looks similar with some UTF handling in the middle

```cpp
  void call(
      out_type<Varchar>& result,
      const arg_type<Varchar>& stringInput,
      const arg_type<Varchar>& pattern,
      const arg_type<Varchar>& replace,
      const arg_type<int64_t>& position) {
    if (performChecks(result, stringInput, pattern, replace, position - 1)) {
      return;
    }
    size_t start = functions::stringImpl::cappedUnicodeBytes(
        stringInput, position - 1);
    if (start > stringInput.size() + 1) {
      result = stringInput;
      return;
    }
    performReplace(result, stringInput, pattern, replace, start);
  }
```

Is there a possibility of leveraging Zero-copy string results here? Sometimes we just copy stringInput or replace into result. My understanding is that because we do not always copy and because there are 2 different args we may copy from that it isn't possible

mbasmanova commented:
> My understanding is that because we do not always copy and because there are 2 different args we may copy from that it isn't possible

I'm thinking the same.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -135,79 +127,101 @@ struct RegexpReplaceFunction {
       const arg_type<Varchar>& replace,
       const arg_type<int64_t>& position) {
-    VELOX_USER_CHECK_GE(position, 1, "regexp_replace requires a position >= 1");
+    if (performChecks(result, stringInput, pattern, replace, position - 1)) {
```
A conventional way is to define a template method with bool isAscii template parameter, then use `functions::stringImpl::cappedUnicode<isAscii>`.

codyschierbeck commented:
@kgpai Can you take a look, with the merging of #8637, this PR is now ready for review.

kgpai commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -135,79 +127,101 @@ struct RegexpReplaceFunction {
     std::string targetString(
-        targetStringPiece.data(), targetStringPiece.size());
-    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+        stringInput.data() + position, stringInput.size() - position);
```
nit: I am not sure about this, but is stringInput.size() size in bytes or unicode code points ? (The arithmetic to compute the targetString might not be correct in former case).

codyschierbeck replied:
In this case, the position is dependent on the function that calls performReplace.

For example, in callAscii the position is always just the user passed position translated to a 0-indexed value:

```cpp
  void callAscii(
      out_type<Varchar>& result,
      const arg_type<Varchar>& stringInput,
      const arg_type<Varchar>& pattern,
      const arg_type<Varchar>& replace) {
    callAscii(result, stringInput, pattern, replace, 1);
  }

  void callAscii(
      out_type<Varchar>& result,
      const arg_type<Varchar>& stringInput,
      const arg_type<Varchar>& pattern,
      const arg_type<Varchar>& replace,
      const arg_type<int64_t>& position) {
    if (performChecks(result, stringInput, pattern, replace, position - 1)) {
      return;
    }
    performReplace(result, stringInput, pattern, replace, position - 1);
  }
```
In this case a character is always a single byte.

In the UTF-8 case, we convert the passed position to bytes in this block:

```cpp
    size_t start = functions::stringImpl::cappedByteLength<false>(
        stringInput, position - 1);
    if (start > stringInput.size() + 1) {
      result = stringInput;
      return;
    }
    performReplace(result, stringInput, pattern, replace, start);
```

kgpai commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -325,11 +331,15 @@ TEST_F(RegexFunctionsTest, regexpReplacePosition) {
 }
 
 TEST_F(RegexFunctionsTest, regexpReplaceNegativePosition) {
```
nit: Can you add a test where you have bad unicode chars in the string and ensure behavior is similar to spark?

kgpai reviewed (commented):
Looks good - some minor nits.

FelixYBW commented:
@PHILO-HE 
There is some pattern mismatch between re2 library and spark. So we need to check the pattern in the validation stage. Should we enumerate the differences in function makeRLike?

PHILO-HE commented:
> @PHILO-HE There is some pattern mismatch between re2 library and spark. So we need to check the pattern in the validation stage. Should we enumerate the differences in function makeRLike?

@FelixYBW, yes, I think so. `makeRlike` calls `ensureRegexIsConstantAndCompatible` which will throw exception for incompatible pattern and make validation process fail for Gluten. Maybe, `regexp_replace` should do the same thing. I can help do that in a separate PR. cc @codyschierbeck

codyschierbeck commented:
@PHILO-HE @FelixYBW @kgpai  I've spent the past few days attempting to validate the differences between Java regex engine and RE2/Boost (PCRE2 based). I used https://github.com/google/re2/wiki/Syntax as a source for the types of patterns possible.

Our checkForCompatiblePattern function blocks more than just character class union, intersection, and differences. It simply looks for nested opening brackets and throws an exception if caught. This will also throw an exception for ascii character classes. For now, I've adjusted the error message and added this to the documentation.

codyschierbeck commented:
@kgpai This PR is ready for review

PHILO-HE commented:
@codyschierbeck, thanks for investigating the incompatible regrex patterns! I think there may also exist a few incompatible cases between boost and java regex. Maybe, RE2 is acceptable for most real-world use cases.

This patch looks good to me. @mbasmanova, could you take a review further? Thanks!

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -38,7 +38,12 @@ See https://github.com/google/re2/wiki/Syntax for more information.
 
     Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite``. If no match is found, the original string is returned as is.
-    There is a limit to the number of unique regexes to be compiled per function call, which is 20.
+    There is a limit to the number of unique regexes to be compiled per function call, which is 20. If this limit is exceeded the function will throw an exception.
```
This documentation applies to all regex functions. Instead repeating it for each function, perhaps, move it to the top of this document.

codyschierbeck replied:
As of now, regexp_extract and rlike call the "ensureRegexIsContstantAndCompatible" function. In that case, would they not be limited to a single regex? I may be unaware, but I am not sure of a use case where they would encounter this exception. I wrote up a quick test for this, results displayed below.
<img width="354" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/c34f5201-c69a-473d-8d58-0063f35d9aa2">
<img width="278" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/b58adcdf-e0e6-41aa-9af8-4e63a32ef833">
<img width="350" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/1146e9e6-4dcd-4594-940b-900d5ac4da6e">

mbasmanova replied:
> would they not be limited to a single regex?

I see. Some functions support only constant regex and other support variable regex (up to 20 unique). Let's make sure this is clear in the docs.

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -38,7 +38,12 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+
+    regexp_replace will throw an exception if ``string`` contains an invalid UTF-8 character, or if ``pattern`` does not conform to RE2 syntax: https://github.com/google/re2/wiki/Syntax.
+
+    regexp_replace does not support named ASCII character classes or ASCII character class union, intersection, or difference and will throw an exception if they are detected within the provided ``pattern``.
```
Is this a limitation of RE2 or this function? Would users understand what this means? Maybe we could provide some links that explain class union, intersection,... or give some examples?

codyschierbeck replied:
This is a limitation of a mixture of RE2 and the way regexp_extract, rlike, and regexp_replace choose to handle that limitation.

regexp_extract and rlike call "ensureRegexIsConstantAndCompatible" which calls checkForCompatiblePattern after ensuring the pattern is constant. regexp_replace skips the constant check and directly calls checkForCompatiblePattern, which is designed to get around a limitation of RE2. 

RE2 does not allow for character class unions, intersections, or differences whereas Java does. This link can help illustrate this: https://regex101.com/r/WiVEbV/1

If you view with the "Java 8" flavor, the output should look like the following:
<img width="444" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/24e0e326-8970-43ed-ab6c-1f4014cf1bd4">
The first line is an input that is designed around Java's regex capabilities. The second line is designed around RE2. RE2, instead of viewing this as a union of character classes "a" and "bc" views it as a single character class of "a[bc" and expects a single "]" at the end. When you switch to RE2's flavor (PCRE2) you see it fully matching the second line:
<img width="434" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/f61c41fc-f176-478a-987d-d777af49cb02">


ASCII character classes (for example [[:alnum::]] ), compile and work correctly in a standalone RE2 environment, but when passed into Velox they fail the checks in checkForCompatiblePattern. 

This is because checkForCompatiblePattern was designed around basic character classes without ASCII character classes in mind. It checks for a nested opening brace and if it detects one throws an error. As shown here, ASCII character classes must contain a nested opening brace:

<img width="377" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/e8a0cf97-4dd7-40ee-a030-b237a32d2aba">

I attempted to modify the function to handle this, but as I got further and further more corner cases became clear to me and I decided that it would probably be best to just document the limitation and allow Gluten to fallback to Java instead.

In regexp_extract and rlike implementations, this limitation is only documented in the error message. I can move these warnings in the regexp_replace documentation towards to top of the file to make it clear they apply to all velox spark regexp functions.

Based on a quick search, it seems Presto also uses the Java pattern syntax with a few limitations: https://prestodb.io/docs/current/functions/regexp.html. I would assume, then, that these limitations also occur when using RE2 as a proxy for a Java regex engine. I have no experience with Presto and have nothing setup locally to check discrepancies, but taking the example function call of regexp_replace `(SELECT regexp_replace('1a 2b 14m', '(\d+)([ab]) ', '3c$2 '); -- '3ca 3cb 14m')` and replacing ([ab]) with a union, we get the following:


<img width="701" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/3bd37ec9-43e6-4506-8e04-8a786ae7f2d3">

Line 1 is the documented case working as expected (though the Velox Presto function documentation does not contain the \\$, just the $).
Line 2 is the documented case using character class union and failing.
Line 3 is the documented case adding the ] to match the way RE2 interprets the pattern and working as "expected".

Seeing the $2 in the Presto use case made me realize that that may be something that I am not accounting for in my implementation. Additionally, after rebasing I've noticed that all prod spark regexp functions implementations are now leveraging implementations in velox/functions/lib. Do you believe I should modify my implementation to do the same?

mbasmanova replied:
> checkForCompatiblePattern, which is designed to get around a limitation of RE2.

I see. I feel it is very difficult and practically impossible to check the regex pattern and decide whether it will be handled by RE2 exactly the same as by Java Regex. Hence, I wonder if we should even try. I'd consider not trying and simply compiling regex and if compilation fails throw an error, otherwise proceed. I do understand that this means that query results with Gluten may be different from Spark Java, but I don't see any way to robustly protect from them, hence, suggest to clearly document that Gluten uses RE2 engine which has different syntax and semantics than Java regex.

codyschierbeck replied:
I agree. 

In regards to "should we even try", I spent about 2 days writing up a software suite that automates these checks on a compile and output matching basis. There are 200+ pattern types in the RE2 syntax page. Of those 200, RE2 as a direct proxy for Java regex only lacks support for ~20 patterns that Java does support.  If instead of throwing errors when a pattern fails to compile, you substitute boost, then the number drops to ~5. But, based on this [page](https://lh3lh3.users.sourceforge.net/reb.shtml) that I've tested locally, the performance degradation when using boost might not be worth it.

codyschierbeck replied:
Do you think we should remove the "checkForCompatiblePattern" and just clearly document the risk of differing output using Velox?

mbasmanova replied:
I think so.

mbasmanova replied:
> I spent about 2 days writing up a software suite that automates these checks on a compile and output matching basis

It might be helpful to summarize your findings in a blog post. CC: @zacw7 @amitkdutta

codyschierbeck replied:
Yeah, I'd be very interested in that. I need to go back and re-write some of it as I never planned to use it after I was done with it, so I wrote some parts directly into the Velox RegexFunctionsTest test suite (which is why I discovered the ASCII character class issue, since they do compile in the standalone environment). I deleted those when I was done in order to make a push to this PR. 

But yeah, once this PR is done I'd be glad to talk with Zac and Amit about my findings.

mbasmanova commented on velox/expression/tests/SparkExpressionFuzzerTest.cpp:
```diff
@@ -45,8 +45,10 @@ int main(int argc, char** argv) {
   // The following list are the Spark UDFs that hit issues
   // For rlike you need the following combo in the only list:
   // rlike, md5 and upper
+  // regexp_replace: https://github.com/facebookincubator/velox/issues/8438
```
Is this for regexp_replace or all regexp_xxx functions? Move the comment right above regexp_xxx (L50) and remove "regexp_replace:"

```
  std::unordered_set<std::string> skipFunctions = {
      "regexp_extract",
      // https://github.com/facebookincubator/velox/issues/8438
      "regexp_replace",
```

codyschierbeck replied:
It's just for regexp_replace. As far as I can tell, the other regexp functions do not leverage the cache issues we discovered in #8438 so it's not relevant to them. Will move comment

mbasmanova replied:
> the other regexp functions do not leverage

Is this intentional? We need to make sure no function is compiling regex per row.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -255,6 +260,18 @@ TEST_F(RegexFunctionsTest, regexpReplaceSimple) {
+  std::string badHalf = "\xF0\x82";
+  VELOX_ASSERT_THROW(
+      testingRegexpReplaceRows({badUTF}, {badHalf}, {"Bad"}), "invalid UTF-8");
+  // python converts above values to below and completes regexp_replace
```
I'm not sure I understand this comment. Why is it taking about Python?

codyschierbeck replied:
This is an oversight on my end. Originally, this test case was a sanity checking test case to model behavior I observed in PySpark/python where I incorrectly noticed that PySpark _seemed_ to replace some bad UTF-8 values with parsable ones. Shown below

<img width="357" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/7a05e6a4-f771-460c-b084-47f4c68932ae">
<img width="155" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/957cfbec-8150-415a-8db7-0bb159e4c623">

As you can see in the byte output, it does not replace them but rather it just does not display \x82. 

I will change this test case to just make sure we throw an error when we encounter bad UTF-8 characters. RE2 simply does not allow these patterns to be compiled so a fallback to vanilla spark is needed as far as I can tell.

mbasmanova reviewed (approved):
@codyschierbeck Looks good to me % some comments.

Yuhta commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -34,6 +35,10 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
   void SetUp() override {
     SparkFunctionBaseTest::SetUp();
     registerRegexpReplace("");
```
You can remove this line now

codyschierbeck replied:
Hi, it's unclear to me which line you are talking about. Are you referring to the GTestUtils.h include? Or the comment and options_.parseIntegerAsBigint?

Yuhta replied:
It's this line `registerRegexpReplace("");`

codyschierbeck replied:
Ah, good catch. Removing, thank you!

codyschierbeck commented:
@mbasmanova 

I've implemented most of the requested review changes, and the ones I did not were discussed above. 
1) ensureRegexIsConstantAndCompatible is now just ensureRegexIsConstant.
2) ensureRegexIsComptabile no longer exists and all references and tests around it have been removed
2) The badUTF8 test cases no longer is about my misinterpreted python use case and instead tests a variety of bad UTF8 inputs
3) I've modified regexp.rst in a variety of places to both explain one example of Java and RE2 divergences, as well as describe the constant pattern requirement of regexp_extract and rlike. I also removed parts of regexp_replace documentation that are covered in the introductory block of that file


I noticed a weird indentation error in the documentation of the second function signature of regexp_extract. This issue persists when I stash my changes and recompile the documentation. It goes away when I remove :noindex:, but my current understanding is that line is necessary.

<img width="686" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/6a004c8c-9bb7-4e48-ab90-07988ee29b10">

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -5,20 +5,34 @@ Regular Expression Functions
 supports only a subset of PCRE syntax and in particular does not support
 backtracking and associated features (e.g. back references).
+Java and RE2 regex output can diverage and users should be cautious that
+the patterns they are using perform similarlly between RE2 and Java.
```
typo in "similarlly"

mbasmanova commented:
@codyschierbeck Thank you, Cody. Such a nice thorough work. I'll get a stamp from @Yuhta and work on merging this.

I saw similar indentation issue before, but can't remember how I solved it.

mbasmanova commented:
@codyschierbeck Cody, would you rebase onto latest main?

codyschierbeck commented:
@mbasmanova Sorry, had a busy morning. Fixing typo and rebasing to main
